File: datasets/C++/Gemini/DataStructures/easy5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>

using namespace std;

struct Node {
    int data;
    Node *left;
    Node *right;
    Node(int d) {
        data = d;
        left = NULL;
        right = NULL;
    }
};

Node* buildTree(const vector<int>& nodes, const vector<tuple<int, int, char>>& edges) {
    if (nodes.empty()) return NULL;
    map<int, Node*> nodeMap;
    for (int val : nodes) {
        nodeMap[val] = new Node(val);
    }

    Node* root = nodeMap[nodes[0]];

    for (const auto& edge : edges) {
        int parentVal = get<0>(edge);
        int childVal = get<1>(edge);
        char dir = get<2>(edge);

        if (nodeMap.find(parentVal) != nodeMap.end() && nodeMap.find(childVal) != nodeMap.end()) {
            Node* parent = nodeMap[parentVal];
            Node* child = nodeMap[childVal];
            if (dir == 'L') {
                parent->left = child;
            } else {
                parent->right = child;
            }
        }
    }
    return root;
}

void topView(Node *root) {
    if (root == NULL) return;

    map<int, int> hdMap; // Horizontal distance -> Node data
    queue<pair<Node*, int>> q;
    q.push({root, 0});

    while (!q.empty()) {
        pair<Node*, int> temp = q.front();
        q.pop();
        Node* curr = temp.first;
        int hd = temp.second;

        // If this is the first node we've seen at this horizontal distance, store it.
        if (hdMap.find(hd) == hdMap.end()) {
            hdMap[hd] = curr->data;
        }

        if (curr->left != NULL) {
            q.push({curr->left, hd - 1});
        }
        if (curr->right != NULL) {
            q.push({curr->right, hd + 1});
        }
    }

    for (auto const& pair : hdMap) {
        cout << pair.second << " ";
    }
    cout << endl;
}

int main() {
    int n;
    cin >> n;
    vector<int> nodes(n);
    for (int i = 0; i < n; ++i) {
        cin >> nodes[i];
    }

    int e;
    cin >> e;
    vector<tuple<int, int, char>> edges(e);
    for (int i = 0; i < e; ++i) {
        int parent, child;
        char dir;
        cin >> parent >> child >> dir;
        edges[i] = make_tuple(parent, child, dir);
    }

    Node* root = buildTree(nodes, edges);
    topView(root);

    return 0;
}

File: datasets/Python/Gemini/DataStructures/easy5.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def build_tree(nodes, edges):
    if not nodes:
        return None
    
    node_map = {val: Node(val) for val in nodes}
    
    # Assuming the first node in the list is the root
    root = node_map[nodes[0]]
    
    for parent, child, direction in edges:
        parent_node = node_map.get(parent)
        child_node = node_map.get(child)
        
        if parent_node and child_node:
            if direction == 'L':
                parent_node.left = child_node
            elif direction == 'R':
                parent_node.right = child_node
    
    return root

def top_view(root):
    if not root:
        return
    
    queue = deque([(root, 0)]) # (node, horizontal_distance)
    hd_map = {}
    
    while queue:
        node, hd = queue.popleft()
        
        # If this horizontal distance is seen for the first time, add it to the map
        if hd not in hd_map:
            hd_map[hd] = node.data
            
        if node.left:
            queue.append((node.left, hd - 1))
        
        if node.right:
            queue.append((node.right, hd + 1))
    
    # Sort the horizontal distances to print in order
    sorted_hds = sorted(hd_map.keys())
    
    result = [hd_map[hd] for hd in sorted_hds]
    print(*result)

def main():
    try:
        n_str = input()
        if not n_str.strip():
            n = 0
        else:
            n = int(n_str)
        
        if n == 0:
            print()
            return

        nodes = list(map(int, input().split()))
        
        e = int(input())
        
        edges = []
        for _ in range(e):
            parent, child, direction = input().split()
            edges.append((int(parent), int(child), direction))
            
        root = build_tree(nodes, edges)
        top_view(root)
    except (IOError, ValueError):
        # Handles cases with no nodes and other parsing errors gracefully
        print()

if __name__ == "__main__":
    main()